### PR TITLE
Allow KServe to have its own local gateways for Serverless mode

### PR DESCRIPTION
**Pull request by israel-hdez**
**What this PR does / why we need it**:

These changes introduce the possibility to configure KServe with its own Istio local gateway, to partially decouple KServe from the Knative local gateway.

Typically, it is OK to re-use the already configured Knative local gateway for KServe uses (as long as configs do not conflict). However, there are cases where having a dedicated local gateway for KServe is beneficial. Just to give some examples:
* To have the ability to use strict mTLS in Istio
* To reduce some pressure on the Knative local gateway by having a dedicated gateway deployment (it still would hit Knative gateway, but only once, rather than twice)
* To be able to configure TLS on cluster-local hostnames (Knative support is still experimental)

To have a dedicated Gateway in KServe, similar configurations to Knative are need to be done. At the very least, and if not having a dedicated gateway deployment, a v1/Service and an Istio Gateway resource need to be created for KServe. Such resources would need to be configured in _localGateway_ and _localGatewayService_. KServe still needs to rely on Knative routing for the KSVCs it creates. Thus, after handling an incoming request and resolving its target, it needs to be forwarded to be handled by Knative. This is the reason for introducing a new `knativeLocalGatewayService` in the ConfigMap.

The removed `ingressService` seems to be unused. Apparently, it became unused when the v1alpha2 API of the InferenceServices was deprecated and removed.

**Type of changes**
Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

**TODO**

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note
Allow KServe to have its own local gateways for Serverless mode
```

### CODE REVIEW

File: charts/kserve-resources/templates/configmap.yaml
Comment (lizzzcai) [inline charts/kserve-resources/templates/configmap.yaml:209]: why not point to the `knative-local-gateway.istio-system.svc.cluster.local` as well, like localGatewayService?

File: charts/kserve-resources/templates/configmap.yaml
Comment (lizzzcai) [inline charts/kserve-resources/templates/configmap.yaml:207]: why need to point to knative's local gateway here even if there is a dedicated gateway for KServe?
I saw KNative support [non-default local gateway](https://knative.dev/docs/install/installing-istio/#updating-the-config-istio-configmap-to-use-a-non-default-local-gateway) and seems like it can support multiple entries? if we updated the dedicated KServe localgateway here as well, do we still need to point to knative local gateway?

**Comment by lizzzcai** [comment]
just something come up in my mind maybe can be a future enhancement if it is useful. 
Do you think having a dedicated gateway CRD to manage the gateway creation and configuration here will be useful? so that users don't need to look into how to configure Istio and KNative, and maybe will be helpful for the raw deployment mode (to support different gateway implementations?)

File: charts/kserve-resources/templates/configmap.yaml
Comment (israel-hdez) [reply charts/kserve-resources/templates/configmap.yaml:207]: It is related to the top-level virtual service that KServe creates. There is some explanation here: https://kserve.github.io/website/latest/admin/serverless/servicemesh/#disable-top-level-virtual-service

File: charts/kserve-resources/templates/configmap.yaml
Comment (israel-hdez) [reply charts/kserve-resources/templates/configmap.yaml:209]: I can do it in this part, as this is just an example, to be explicit about what would be the default value.

Although, I'm not sure if it is OK to be explicit in the actual configs. I'm not sure if that would lead to configuration mistakes for users that would be migrating and using a different name for the gateways. I'm trying to be backwards-compat friendly.

File: config/configmap/inferenceservice.yaml
Comment (spolti) [inline config/configmap/inferenceservice.yaml]:
```suggestion
           # knativeLocalGatewayService specifies the hostname of the Knative's local gateway service.
```

File: charts/kserve-resources/templates/configmap.yaml
Comment (spolti) [inline charts/kserve-resources/templates/configmap.yaml]:
```suggestion
           # knativeLocalGatewayService specifies the hostname of the Knative's local gateway service.
```

**Review by spolti — APPROVED** [review]

**Comment by israel-hdez** [comment]
I have a pull request in the website that may give better idea about why these changes are relevant: https://github.com/kserve/website/pull/374.

**Comment by israel-hdez** [comment]
> just something come up in my mind maybe can be a future enhancement if it is useful. Do you think having a dedicated gateway CRD to manage the gateway creation and configuration here will be useful? so that users don't need to look into how to configure Istio and KNative, and maybe will be helpful for the raw deployment mode (to support different gateway implementations?)

Wel... I like abstractions to make things easier for users. But, unfortunately, there are many ways to expose services and it is probably better and less problematic to let the users do it themselves using their favorite method.

That said, I expect [Gateway API](https://gateway-api.sigs.k8s.io/) will be that needed abstraction.

**Review by spolti — APPROVED** [review]
/lgtm

File: charts/kserve-resources/templates/configmap.yaml
Comment (israel-hdez) [reply charts/kserve-resources/templates/configmap.yaml]: oops! Fixed.

File: config/configmap/inferenceservice.yaml
Comment (israel-hdez) [reply config/configmap/inferenceservice.yaml]: Fixed :-)

**Comment by israel-hdez** [comment]
@yuzliu I have an approval. Would you like to review?

**Comment by israel-hdez** [comment]
Sorry... in my previous comment I wanted to tag @yuzisun

File: charts/kserve-resources/templates/configmap.yaml
Comment (yuzisun) [inline charts/kserve-resources/templates/configmap.yaml:203]: This is not backwards compatible if user has customized the existing `localGatewayService`

Comment (yuzisun) [reply charts/kserve-resources/templates/configmap.yaml:203]: oh I see you have
```go
if len(ingressConfig.KnativeLocalGatewayService) == 0 {
     ingressConfig.KnativeLocalGatewayService = ingressConfig.LocalGatewayServiceName
}
```

File: charts/kserve-resources/templates/configmap.yaml
Comment (yuzisun) [reply charts/kserve-resources/templates/configmap.yaml:207]: @israel-hdez Can you explain a bit more how this works when top level virtual service is enabled by using an additional gateway? I am a bit concerned that this adds additional configuration complexity to kserve admin.

Comment (yuzisun) [reply charts/kserve-resources/templates/configmap.yaml:207]: If I understand correctly, the additional local gateway is needed for handling the non service mesh traffic when mutual tls is on. But I am not sure if this is a real use case as if the cluster is turned on for service mesh then mutual tls should work in cluster. You can use ingress gateway to handle the non service mesh traffic so we eliminate the need of the additional local gateway.

**Comment by yuzisun** [comment]
@israel-hdez @terrytangyuan I am not convinced we need this additional configuration but I could miss something, maybe we can schedule a call to discuss.

**Comment by israel-hdez** [comment]
@yuzisun We'll reach you on Slack either on Monday 8th or Tuesday 9th to schedule the call.

File: charts/kserve-resources/templates/configmap.yaml
Comment (yuzisun) [inline charts/kserve-resources/templates/configmap.yaml:184]: just a note that this config is no longer used by kserve hence we can remove it

File: charts/kserve-resources/templates/configmap.yaml
Comment (yuzisun) [reply charts/kserve-resources/templates/configmap.yaml:209]: I am just thinking about a better name for this as by default other gateway configurations are reusing knative gateways, so this could lead to some confusions.

File: charts/kserve-resources/templates/configmap.yaml
Comment (israel-hdez) [reply charts/kserve-resources/templates/configmap.yaml:203]: Yes, the code is what makes it backwards compatible.

File: charts/kserve-resources/templates/configmap.yaml
Comment (israel-hdez) [reply charts/kserve-resources/templates/configmap.yaml:209]: The way I think it is that, by default, _kserve gateway_ is also the Knative's one :slightly_smiling_face: .

I was unable to think a better name, given in the de-coupled configuration, this name seems to represent its goal. But if there is a better naming, I'm happy to change it.

File: charts/kserve-resources/templates/configmap.yaml
Comment (israel-hdez) [reply charts/kserve-resources/templates/configmap.yaml:207]: > But I am not sure if this is a real use case as if the cluster is turned on for service mesh then mutual tls should work in cluster.

@yuzisun  Hopefully, it is more clear after our offline call.

For the record, and in short, when locking the mesh with mTLS using both _PeerAuthentication_ and also _DestinationRule_ traffic becomes blocked on the ingress-gateway when using the top-level virtual service. This is because the top-level virtual service is configuring a loop on the ingress gateway. The listening port of the ingress gateway is using plain-text. In the default configurations, this loopback connection of the top-level virtualservice is possible, because the connection happens as plain-text. However, when locking the mesh with a DestinationRule with mTLS, this loopback connection no longer succeeds because it is tried with mTLS while the gateway is still listening with plain-text.

By allowing KServe to have its own Gateway, it becomes possible to open a new port in the gateway that would listen with mTLS rather than plain-text. Thus, it becomes feasible to configure the top-level virtualservice to connect to the mTLS port.

**Comment by israel-hdez** [comment]
@yuzisun After some long vacations, I'm retaking this. I wonder if there is something else that needs to be clarified for this PR?

**Comment by yuzisun** [comment]
> @yuzisun After some long vacations, I'm retaking this. I wonder if there is something else that needs to be clarified for this PR?

other than the configuration naming it looks good

**Review by yuzisun — APPROVED** [review]

**Pull request merged**
